src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -802,6 +828,56 @@ public void Close()
                 _terminateDelegate(_nodeId);
             }
 
+            /// <summary>
+            /// Waits for the child node process to exit.
+            /// </summary>
+            public async Task WaitForExitAsync()
+            {
+                int processId = _processId;
+                if (processId != -1)
+                {
+                    Process childProcess;
+                    try
+                    {
+                        childProcess = Process.GetProcessById(processId);
+                    }
+                    catch (System.ArgumentException)
+                    {
+                        // The process has terminated already.
+                        return;
+                    }
+
+                    // Wait for the process to terminate.
+                    CommunicationsUtilities.Trace("Waiting for node with pid = {0} to terminate", processId);
+
+                    if (_closeSent)
+                    {
+                        // .NET 5 introduces a real WaitForExitAsyc.
+                        // This is a poor man's implementation that uses polling.
+                        int timeout = TimeoutForWaitForExit;
+                        int delay = 5;
+                        while (timeout > 0)
+                        {
+                            bool exited = childProcess.WaitForExit(milliseconds: 0);
+                            if (exited)
+                            {
+                                return;
+                            }
+                            timeout -= delay;
+                            await Task.Delay(delay).ConfigureAwait(false);
+
+                            // Double delay up to 500ms.
+                            delay = Math.Min(delay * 2, 500);
+                        }
+                    }
+
+                    // Kill the child and do a blocking wait.
+                    CommunicationsUtilities.Trace("Killing node with pid = {0}", processId);
+                    childProcess.Kill();
```

Comment:
I think rainersigwald told me it was bad to kill child processes explicitly because it left our logging incomplete and could leave VS in a bad state when starting up again. The latter doesn't matter for command line builds, clearly.

Also, if we're planning to kill it explicitly, NativeMethodsShared.KillTree does that.